### PR TITLE
enhance memory optimization transpiler to support user defined skip_opt_set

### LINKED ISSUE #11320: 启动内存优化后，多个loss训练，获取loss值有异常
我的多个loss的程序调用fluid.memory_optimize(fluid.default_main_program())内存优化后，获取loss显示异常。并且相同环境，关闭fluid.memory_optimize优化，两个loss值有diff。
image
我调用的一些关键程序：
train_exe = fluid.ParallelExecutor(use_cuda=True, loss_name=avg_loss.name)
fetch_list_var = []
results = train_exe.run(fetch_list_var,  feed=feed_dict)

建议可以拿googlenet的训练试试，看看是否可以复现。

@panyx0718

### PR DESCRIPTION
**Pull request by QiJune**
Try to fix #11320

### CODE REVIEW

File: python/paddle/fluid/transpiler/memory_optimization_transpiler.py [hunk 1]
```diff
@@ -358,7 +363,7 @@ def _get_cfgs(input_program):
     return cfgs
 
 
-def memory_optimize(input_program, print_log=False, level=0):
+def memory_optimize(input_program, skip_opt_set=None, print_log=False, level=0):
```
Comment (qingqing01) [inline python/paddle/fluid/transpiler/memory_optimization_transpiler.py]: 1. Need to add comments for the new arguments.
2. Need unit test.

Comment (zchengquan) [reply python/paddle/fluid/transpiler/memory_optimization_transpiler.py]: 直接修改本地安装环境中对应文件，进行测试，获取的loss值依旧异常

Comment (QiJune) [reply python/paddle/fluid/transpiler/memory_optimization_transpiler.py]: @zchengquan 你需要从源码编译安装
在修改对应python文件后，请重新执行make install

**Review by dzhwinter — APPROVED** [review]
Great PR, had help with @dyning fix this issue. I will do polish the api and add more tests in https://github.com/PaddlePaddle/Paddle/pull/11462

**Pull request merged**